component/prometheus/source/api/api.go

```diff
@@ -0,0 +1,179 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+	"sync"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	agentprom "github.com/grafana/agent/component/prometheus"
+	"github.com/grafana/agent/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/weaveworks/common/logging"
+	"github.com/weaveworks/common/server"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "prometheus.source.api",
+		Args: Arguments{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+type Arguments struct {
+	HTTPAddress string               `river:"http_address,attr"`
+	HTTPPort    int                  `river:"http_port,attr"`
+	ForwardTo   []storage.Appendable `river:"forward_to,attr"`
+}
+
+type Component struct {
+	opts               component.Options
+	handler            http.Handler
+	serverErr          chan error
+	fanout             *agentprom.Fanout
+	uncheckedCollector *util.UncheckedCollector
+
+	updateMut sync.RWMutex
+	args      Arguments
+	server    *server.Server
+}
+
+func New(opts component.Options, args Arguments) (component.Component, error) {
+	fanout := agentprom.NewFanout(args.ForwardTo, opts.ID, opts.Registerer)
+	uncheckedCollector := util.NewUncheckedCollector(nil)
+	c := &Component{
+		opts:               opts,
+		handler:            remote.NewWriteHandler(opts.Logger, fanout),
+		serverErr:          make(chan error),
+		fanout:             fanout,
+		args:               args,
+		uncheckedCollector: uncheckedCollector,
+	}
+	opts.Registerer.MustRegister(uncheckedCollector)
+
+	// we do not need to hold the lock here, since `c` is not yet exposed
+	err := c.createNewServer(args)
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	go c.runServer()
+	defer func() {
+		c.updateMut.Lock()
+		defer c.updateMut.Unlock()
+		c.shutdownServer()
+	}()
+
+	for {
+		select {
+		case err := <-c.serverErr:
+			return fmt.Errorf("server terminated with error: %v", err)
```

Comment:
We don't recommend having any kind of signal which would exit the Run function, that's effectively fatal for the component; nothing will cause it to rerun. 

Instead, if there's an error like this, expose the component as unhealthy by implementing [component.HealthComponent](https://github.com/grafana/agent/blob/93cb1a2718f6fc90fd06ef33b6bcff519dbed662/component/component_health.go#L14), and allow the component to return to a healthy state once the user reconfigures the component or triggers a reload.

Reply:
Thanks for spotting this and the suggestion. I've implemented this now :)